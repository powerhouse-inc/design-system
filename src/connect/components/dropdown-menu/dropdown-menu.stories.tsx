import type { Meta, StoryObj } from '@storybook/react';
import { ConnectDropdownMenu } from './dropdown-menu';

import { Icon } from '@/powerhouse';
import { useState } from 'react';

const meta = {
    title: 'Connect/Components/DropdownMenu',
    component: ConnectDropdownMenu,
    argTypes: {
        onItemClick: { action: 'onItemClick' },
        items: { control: { type: 'object' } },
        className: { control: { type: 'text' } },
        menuClassName: { control: { type: 'text' } },
        menuItemClassName: { control: { type: 'text' } },
    },
} satisfies Meta<typeof ConnectDropdownMenu>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    render: function ControlledWrapper(args) {
        const [isOpen, setIsOpen] = useState(false);
        function onOpenChange() {
            setIsOpen(!isOpen);
        }

        return (
            <div>
                <button onClick={onOpenChange}>Toggle</button>
                <ConnectDropdownMenu
                    {...args}
                    isOpen={isOpen}
                    onOpenChange={onOpenChange}
                />
            </div>
        );
    },
    args: {
        isOpen: false,
        onOpenChange: () => {},
        onItemClick: () => {},
        className:
            'bg-blue-500 text-white w-8 h-8 rounded justify-center items-center flex',
        menuClassName: 'bg-white cursor-pointer',
        menuItemClassName: 'hover:bg-slate-50 px-2',
        items: [
            {
                id: 'duplicate',
                label: 'Duplicate',
                icon: <Icon name="files-earmark" />,
            },
            {
                id: 'new-folder',
                label: 'New Folder',
                icon: <Icon name="folder-plus" />,
            },
            {
                id: 'rename',
                label: 'Rename',
                icon: <Icon name="pencil" />,
            },
            {
                id: 'delete',
                label: 'Delete',
                icon: <Icon name="trash" />,
                className: 'text-red-900',
            },
        ],
    },
};
