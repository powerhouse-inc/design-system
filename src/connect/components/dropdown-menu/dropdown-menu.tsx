import { ControlledDropdownMenuProps, DropdownMenu } from '@/powerhouse';
import { useMemo } from 'react';
import { twMerge } from 'tailwind-merge';

export interface ConnectDropdownMenuItem<T extends string = string> {
    id: T;
    icon?: React.JSX.Element;
    label: string;
    className?: string;
}

export interface ConnectDropdownMenuProps<T extends string = string>
    extends Omit<ControlledDropdownMenuProps<T>, 'items'> {
    items: ConnectDropdownMenuItem<T>[];
}

export function ConnectDropdownMenu<T extends string = string>(
    props: ConnectDropdownMenuProps<T>,
) {
    const { items, onItemClick, menuClassName, ...dropDownProps } = props;

    const dropdownItems = useMemo<ControlledDropdownMenuProps<T>['items']>(
        () =>
            items.map(item => ({
                id: item.id,
                content: (
                    <div
                        key={item.id}
                        className={twMerge(
                            'flex h-9 flex-row items-center px-3',
                            item.className,
                        )}
                    >
                        {item.icon && (
                            <span className="mr-2 inline-block">
                                {item.icon}
                            </span>
                        )}
                        <div>{item.label}</div>
                    </div>
                ),
            })),
        [items],
    );

    return (
        <DropdownMenu<T>
            items={dropdownItems}
            onItemClick={onItemClick}
            menuClassName={twMerge(
                'py-3 rounded-2xl modal-shadow text-sm font-medium text-[#6F767E]',
                menuClassName,
            )}
            {...dropDownProps}
        />
    );
}
