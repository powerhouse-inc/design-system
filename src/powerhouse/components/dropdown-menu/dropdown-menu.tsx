import React from 'react';
import {
    Button,
    Item,
    Menu,
    MenuTrigger,
    Popover,
    PopoverProps,
} from 'react-aria-components';
import { twMerge } from 'tailwind-merge';

export interface SharedDropdownMenuProps {
    onItemClick: (itemId: React.Key) => void;
    items: Array<{ id: React.Key; content: React.ReactNode }>;
    className?: string;
    menuClassName?: string;
    menuItemClassName?: string;
    popoverProps?: PopoverProps;
}

export type ControlledDropdownMenuProps = SharedDropdownMenuProps & {
    type: 'controlled';
    isOpen: boolean;
    onOpenChange: (isOpen: boolean) => void;
};

export type UncontrolledDropdownMenuProps = SharedDropdownMenuProps & {
    type: 'uncontrolled';
    children: React.ReactNode;
};

export type DropdownMenuProps =
    | ControlledDropdownMenuProps
    | UncontrolledDropdownMenuProps;

/**
 * A dropdown menu component based on react-aria-components' <Menu />.
 *
 * @see https://react-spectrum.adobe.com/react-aria/Menu.html
 *
 * Supports both controlled and uncontrolled modes.
 *
 * Controlled mode is useful when you want to control the open state of the menu, but you must provide your own state and button implementation to toggle the menu.
 *
 * Uncontrolled mode is useful when you want to use the default button and state provided by the component. In uncontrolled mode, you must pass a `children` prop to the component, which will be used as the content of the button that toggles the menu.
 */
export function DropdownMenu(props: DropdownMenuProps) {
    const {
        items,
        className,
        onItemClick,
        popoverProps,
        menuClassName,
        menuItemClassName,
    } = props;

    const Component = () => (
        <Popover {...popoverProps}>
            <Menu
                onAction={key => onItemClick(key)}
                className={twMerge(
                    'outline-none overflow-hidden',
                    menuClassName,
                )}
            >
                {items.map(item => (
                    <Item
                        id={item.id}
                        key={item.id}
                        className={twMerge('outline-none', menuItemClassName)}
                    >
                        {item.content}
                    </Item>
                ))}
            </Menu>
        </Popover>
    );

    if (props.type === 'controlled') {
        const { isOpen, onOpenChange } = props;
        return (
            <MenuTrigger isOpen={isOpen} onOpenChange={onOpenChange}>
                <Component />
            </MenuTrigger>
        );
    }

    return (
        <MenuTrigger>
            <Button
                aria-label="menu"
                className={twMerge('outline-none', className)}
            >
                {props.children}
            </Button>
            <Component />
        </MenuTrigger>
    );
}
