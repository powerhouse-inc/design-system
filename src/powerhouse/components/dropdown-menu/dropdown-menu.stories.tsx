import type { Meta, StoryObj } from '@storybook/react';
import { DropdownMenu } from './dropdown-menu';

const meta = {
    title: 'Powerhouse/Components/DropdownMenu',
    component: DropdownMenu,
    argTypes: {
        onItemClick: { action: 'onItemClick' },
        children: { control: { type: 'text' } },
        items: { control: { type: 'object' } },
        className: { control: { type: 'text' } },
        menuClassName: { control: { type: 'text' } },
        menuItemClassName: { control: { type: 'text' } },
    },
} satisfies Meta<typeof DropdownMenu>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
    args: {
        className:
            'bg-blue-500 text-white w-8 h-8 rounded justify-center items-center flex',
        menuClassName:
            'border-2 border-indigo-600 w-64 rounded bg-white cursor-pointer',
        menuItemClassName: 'hover:bg-gray-200 px-2',
        children: 'â˜°',
        items: [
            {
                id: 'item-1',
                content: 'Item 1',
            },
            {
                id: 'item-2',
                content: 'Item 2',
            },
            {
                id: 'item-3',
                content: 'Item 3',
            },
        ],
    },
};
