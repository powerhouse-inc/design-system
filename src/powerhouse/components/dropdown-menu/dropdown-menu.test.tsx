import { fireEvent, render, screen } from '@testing-library/react';
import { DropdownMenu, DropdownMenuProps } from './dropdown-menu';

const children = 'â˜°';
const items: DropdownMenuProps['items'] = [
    {
        id: 'item-1',
        content: 'Item 1',
    },
    {
        id: 'item-2',
        content: 'Item 2',
    },
    {
        id: 'item-3',
        content: 'Item 3',
    },
];

describe('DropDownMenu Component', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(
            <DropdownMenu onItemClick={() => {}} items={items}>
                <div>{children}</div>
            </DropdownMenu>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly', () => {
        render(
            <DropdownMenu onItemClick={() => {}} items={items}>
                <div>{children}</div>
            </DropdownMenu>,
        );

        expect(screen.getByText(children)).toBeInTheDocument();
    });

    it('should open menu when click in menu button', () => {
        render(
            <DropdownMenu onItemClick={() => {}} items={items}>
                <div>{children}</div>
            </DropdownMenu>,
        );

        fireEvent(
            screen.getByText(children),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            }),
        );

        expect(
            screen.getByText(items[0].content as string),
        ).toBeInTheDocument();
    });

    it.each(items)('should call fn callback for item %s', item => {
        const onItemClick = jest.fn();

        render(
            <DropdownMenu onItemClick={onItemClick} items={items}>
                <div>{children}</div>
            </DropdownMenu>,
        );

        fireEvent(
            screen.getByText(children),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            }),
        );

        fireEvent(
            screen.getByText(item.content as string),
            new MouseEvent('click', {
                bubbles: true,
                cancelable: true,
            }),
        );

        expect(onItemClick).toHaveBeenCalledWith(item.id);
    });
});
