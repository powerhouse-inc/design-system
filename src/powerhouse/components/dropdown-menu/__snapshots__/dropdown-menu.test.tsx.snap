// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownMenu Component should match snapshot 1`] = `
<DocumentFragment>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="menu"
    class="outline-none"
    data-rac=""
    id="react-aria-:r0:"
    type="button"
  >
    <div>
      â˜°
    </div>
  </button>
</DocumentFragment>
`;
